notebooks/ch16/classification_sgd.md

```diff
@@ -0,0 +1,83 @@
+
+# Stochastic Gradient Descent
+
+Our solution is **stochastic gradient descent**, a technique in which we approximate the true gradient of the cross entropy cost by using just one randomly chosen observation. Since the observation is chosen randomly, we expect that this observation's loss gradient is equal to the true gradient of the cross entropy cost. Our approximation of the gradient of the cross entropy cost is shown below, with $ \sigma_i = f_\hat{\theta}(X_i) = \sigma(X_i \cdot \hat \theta) $.
+
+$$
+\nabla_\theta L(\theta, y) \approx -(y_i - \sigma_i)X_i
+$$
+
+We can then plug this back into the general formula for gradient descent to find estimates for $\theta$:
+
+$$
+\begin{align}
+\theta_{t+1} &= \theta_t - \alpha \cdot \nabla_\theta L(\theta, y) \\
+&\approx \theta_t + \alpha \cdot (y_i - \sigma_i)X_i
+\end{align}
+$$
+
+## Visualizing Stochastic Gradient Descent
+
+Below are visual examples of loss minimization in batch gradient descent and stochastic gradient descent.
+
+![](gd.png)
+![](sgd.png)
+
+At each iteration of batch gradient descent, we move in the direction of the true gradient of the loss (shown by the ellipses). On the other hand, at each iteration of stochastic gradient, our steps may not lead us in the direction of the true gradient; however, our $\theta$ estimates eventually converge to the true loss-minimizing values. At large values of $n$, this is much more desirable due to the fast speed and low computational cost at which we can generate estimates at each iteration.
+
+# Mini-batch Gradient Descent
+
+We can improve our estimates by increasing the number of observations that we select at each iteration. In **mini-batch gradient descent**, we take a simple random sample (known as a mini-batch) of observations and use the average of their loss gradients to construct an estimate of the true gradient of the cross entropy cost. Since our sample is randomly selected, the expectation of this estimate is equal to the true gradient. This is illustrated in the approximation below, where $\mathcal{B}$ is the mini-batch of data points that we randomly sample from the $n$ observations.
+
+$$
+\nabla_\theta L(\theta, y) \approx -\frac{1}{|\mathcal{B}|} \sum_{i\in\mathcal{B}}(y_i - \sigma_i)X_i
+$$
+
+In order to select our mini-batch, we could randomly sample at each iteration. In practice, we randomly shuffle the training data and select our mini-batch by looping through the shuffled data. This also applies to stochastic gradient descent.
+
+
+## Selecting the Mini-Batch Size
+
+The mini-batch size is often set to small numbers such as 1 or 2 to minimize computational cost. However, when working with parallel machines such as a GPU or a distributed computing system, we can increase the mini-batch size to take advantage of the hardware. Since mini-batches can be computed in parallel, the runtime of larger mini-batches is roughly equivalent to the runtime of stochastic gradient descent.
```

Comment:
I *think* mini-batch sizes are often set to somewhere around 10-100 depending on the GPU memory. I don't think we need to mention the case where you run mini-batch without a GPU since mini-batch isn't super useful without one. Instead, we should mention how a relatively small batch size increases the accuracy of the gradient without increasing computation time when using a GPU. (For reference: https://stats.stackexchange.com/questions/164876/tradeoff-batch-size-vs-number-of-iterations-to-train-a-neural-network .)

Also, we should explain the GPU abbreviation before using it.

Reply:
Semi-fixed; is the stackexchange post in the scope of the class?

Reply:
I don't think the stackexchange post is in scope; I just included it for your reference. I think we should just say the batches are often 10-100.